### PR TITLE
Enable charm --version command.

### PR DESCRIPTION
**Pull request by frankban**

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/109/
Test PASSed.

File: cmd/charm/charmcmd/cmd.go [hunk 1]
```diff
@@ -55,6 +55,7 @@ func New() *cmd.SuperCommand {
 			DefaultConfig: os.Getenv(osenv.JujuLoggingConfigEnvKey),
 		},
 		NotifyHelp: notifyHelp,
+		Version:    "2.1.1",
```
Comment (mhilton) [inline cmd/charm/charmcmd/cmd.go]: I can see this getting out of date very quickly. Would it be possible to use the Makefile to set this based on tags like in charmstore?

**Comment by mhilton** [comment]
:-1: on this I'm afraid, I'd prefer not to have to remember to update this every time before releasing.

File: cmd/charm/charmcmd/cmd.go (hunk 1 above)
Comment (rogpeppe) [inline cmd/charm/charmcmd/cmd.go]: Which version is this?

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/110/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/111/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/112/
Test FAILed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/113/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/114/
Test PASSed.

**Comment by mhilton** [comment]
:+1: thanks

**Comment by marcoceppi** [comment]
Does this also create a `charm version` command? If so, I'd like to request
it only be `charm --version`as charm-tools has a version subcommand which
prints both charm and charm-tools versions (used for bug reporting, etc).

**Comment by frankban** [comment]
Well, this implements both and I am not sure about having the flag and the subcommand separated. They could go out of sync, which would be awkward.

**Comment by marcoceppi** [comment]
They won't get out of sync, the charm-version plugin will simply query
`charm --version` for it's portion of the output.

**Comment by frankban** [comment]
I am not sure I like the solution, for a couple of reasons:
1. we'd need to extend the juju/cmd framework to allow this scenario, or just reimplement part of that in the charm command, just for this specific case (juju --version and juju version work as expected);
2. it seems a bit weird that the charm command has a version subcmd only when charm tools are installed, otherwise --version must be used; it could be not straightforward to communicate that to users;
3. I am sure charm tools would implement their own version subcommand so that it uses the output of --version, but what about old charm tools? And what about other possible charm-version plugins? In theory they can still go out of sync, and I have the impression that that's because the charm command should really take ownership of the version name.

Couldn't charm tools just implement their own tools-version command, or similar?

**Comment by marcoceppi** [comment]
I don't think you're framing this marriage of tools appropriately.

Anyone using charmstore-client /will/ have charm-tools installed, except in extraordinary scenarios. Because of how charmstore-client operates now, charm-tools no longer as any user facing central command. So you can't just run `charm-tools --version`. If a `charm version` command is created here, it obfuscates the installed version which, while they are typically updated in lock-step there even today is version drift as changes are made to each independently. charmstore-client in xenial is 2.1.1 and charm-tools is 2.1.2. We're gearing up for a 2.2.0 for Yakkety which I don't think charmstore-client will follow.

If you're dead set on continuing to make troubleshooting user problems more difficult from a developer ecosystem side, I'd like to request code be added that if charm-version is present it's run instead of charmstore-clients implementation. This will, again, query `charm --version` to output the appropriate information when charm-tools is installed alongside charmstore-client.

> Couldn't charm tools just implement their own tools-version command, or similar?

We have already, it's been there for years. You're effectively breaking it, it's `charm version`

Ultimately, you have to remember that charm-tools has owned the charm namespace for well over 4 years now, and is used almost exclusively by charm developers. These store additions are great but they can not come at harming the existing community and commands.

**Comment by frankban** [comment]
I am not suggesting to run `charm-tools --version`, I know it does not make sense. I was suggesting something similar to `charm tools-version` (meaning having a `charm-tools-version` command in the PATH). I am not very interested in knowing who owned the charm namespace and for how long, I am interested in the user perspective, as they'll have `charm --version` and `charm version` behaving differently, and the latter not working if the charm tools are not installed (`go get github.com/juju/charmstore-client`). To me this is like having two different outputs from `charm help` and `charm --help`: that's a WTF moment for both users and developers IMHO. I have raised some points in my previous comment, explaining why in my opinion overloading the version concept is not a good idea and it's not easy to explain to users (as opposed to keeping version and tools-version as two separate concepts and commands), but surely I don't want to make troubleshooting user problems more difficult, so I am pausing this change so that the proper solution can be implemented starting from this branch. Thanks for your feedback!

**Comment by urosj** [comment]
@marcoceppi @frankban As long as the code that is used to build the package gets tagged (as it is already the case), we only need the existing charm version. There is no need for this branch, TBH. "charm version" will show us, the developers, enough information to know from which code base the bug comes.

**Pull request closed without merging**